Implement `TaffyLayout` widget

nicoburns opened the pull request:
# Objective

Enable Taffy based layout to be used with Xilem

## Changes made

- Gitignore .DS_Store files 
- Add `compute_max_intrinsic` method to `Pod` (this method already existed on the `Widget` trait)
- Add `downcast_ref` method to `Pod` (there was already a `downcast_mut` method)
- Added `cx.request_paint()` in the layout method of the button widget (otherwise it wasn't repainting on window resize)
- Create `TaffyLayout` widget and corresponding view (adapted from the `LinearLayout`). Constructors for the view currently include `flex_row`, `flex_column`, and `div`.
- Create `taffy` example which uses the new view/widget.

## Notes

- For the most part this could have been implemented as a 3rd party widget. The only changes required outside of creating the view/widget themselves were minor and not taffy-specific.
- ~This is currently depending on it's own branch of Taffy. But the changes are relatively minor (making Taffy pass along which Axis it is measuring when requesting child measurement), and I don't anticipate major issues getting them merged. (https://github.com/DioxusLabs/taffy/pull/575)~ Changes now merged to Taffy's `main` branch. I would anticipate a few further minor changes, but the integration works as-is.
- Caching is important for performance and is not implemented yet. There are no known blockers, just haven't gotten around to this yet.
- There is scope for the API to be improved. Currently it just stores a whole `taffy::style::Style` on each widget. But it would possible to create more specialised widgets to reduce memory overhead.

## Screenshot

<img width="1458" alt="Screenshot 2023-11-02 at 00 28 22" src="https://github.com/linebender/xilem/assets/1007307/09df1695-f1cc-482e-bc8a-7801adbd7a63">

jaredoconnell reviewed (commented):
Is it possible to reduce the redundancy between layout and compute_max_intrinsic?
Seems like a good start to get stuff moving. The cache will be very important in the future.

nicoburns commented on src/view/taffy_layout.rs:
```diff
@@ -0,0 +1,148 @@
+///
+/// Each Element is positioned on the specified Axis starting at the beginning with the given spacing
+///
+/// This View is only temporary is probably going to be replaced by something like Druid's Flex
```
Ah, that comment was copied from `linear_layout`.

nicoburns commented on Cargo.toml:
```diff
@@ -47,6 +47,7 @@ wayland = ["glazier/wayland"]
 
 [dependencies]
 xilem_core.workspace = true
+taffy = { git = "https://github.com/nicoburns/taffy", rev = "6d7f3cf86a923a02c9d7b67053096a8b3486f15f" }
```
Yes, good spot!

nicoburns replied:
done

giannissc commented:
I will finally get around to reviewing this today!

> Caching is important for performance and is not implemented yet. There are no known blockers, just haven't gotten around to this yet.
Is this something that will be in a separate PR? @nicoburns

nicoburns commented:
> > Caching is important for performance and is not implemented yet. There are no known blockers, just haven't gotten around to this yet.
> 
> Is this something that will be in a separate PR? @nicoburns

I had a quick think about how this might be implemented, and I think it will need to be. Adding the cache is relatively straightforward. The thing I am currently blocked on is cache invalidation. If the styles of a view/widget change, then the cache would need to be invalidated for that widget *and all parents*. And I'm not currently sure how to do that? (perhaps a message / event - but how does one send one of those?). It would also be possible to forgo caching between multiple layout runs (and only have it within a single run) and somehow have a "prelayout" or "postlayout" event in which we drop the caches.

DJMcNab commented on src/view/taffy_layout.rs:
```diff
@@ -0,0 +1,149 @@
+}
+
+/// creates a horizontal [`TaffyLayout`].
+pub fn div<T, A, VT: ViewSequence<T, A>>(children: VT) -> TaffyLayout<T, A, VT> {
```
Div isn't a very compelling name. If there's an alternative you can think of, I'd like to change to it

nicoburns replied:
Hmm... it's called `div` because that's the name on web, and I can't think of why you'd want to use block layout unless you're emulating web. I could just remove it entirely? Or call it something like `raw_taffy_layout`?

DJMcNab commented on src/widget/taffy_layout.rs:
```diff
@@ -0,0 +1,454 @@
+// Copyright 2022 The Druid Authors.
```
The occasional uses of the absolute path to reference `taffy` in this file are very unusual. Is this be referring to a different `taffy` (module vs crate?)? I can't see the module/other import in that case

nicoburns replied:
Ah, this is an artefact of a workaround I was using (an inline `mod taffy`) to workaround the fact that Taffy wasn't exporting most types at the root level. That has now been corrected so I've updated everything to import from `taffy::`.

nicoburns commented on Cargo.toml:
```diff
@@ -47,6 +47,7 @@ wayland = ["glazier/wayland"]
 
 [dependencies]
 xilem_core.workspace = true
+taffy = { git = "https://github.com/DioxusLabs/taffy", rev = "7781c70241f7f572130c13106f2a869a9cf80885" }
```
A `taffy` feature flag has been added. I've made it default enabled for the time being. Obviously that can be changed if desired.

nicoburns commented:
@giannissc 

> > Caching is important for performance and is not implemented yet. There are no known blockers, just haven't gotten around to this yet.

> Is this something that will be in a separate PR?

Caching is now implemented. It really ought to be implemented for the `LinearLayout` and `Text` widgets too, but I think those implementations can be in follow up PRs.

nicoburns commented on examples/taffy.rs:
```diff
@@ -0,0 +1,59 @@
+            s.gap.width = length(20.0);
+            s.flex_grow = 1.0;
+            s.justify_content = Some(JustifyContent::Center);
+            s.align_items = Some(AlignItems::Center);
```
Yeah, I think builder methods would be nice. I figure they can easily be added later though.

nicoburns commented on src/view/taffy_layout.rs:
```diff
@@ -0,0 +1,149 @@
+}
+
+/// creates a horizontal [`TaffyLayout`].
+pub fn flex_row<T, A, VT: ViewSequence<T, A>>(children: VT) -> TaffyLayout<T, A, VT> {
```
No reason! I just didn't add one yet.

nicoburns replied:
There is now a `grid` constructor

nicoburns commented:
> Is it possible to reduce the redundancy between layout and compute_max_intrinsic? Seems like a good start to get stuff moving. The cache will be very important in the future.

Not easily. As things stand you could combine them into one method with a parameter instead. But I would expect these methods to further diverge over time, and IMO this would already be more confusing than a small amount of duplication.

nicoburns commented on src/view/taffy_layout.rs:
```diff
@@ -0,0 +1,149 @@
+    TaffyLayout::new_flex(children, taffy::FlexDirection::Column)
+}
+
+/// creates a horizontal [`TaffyLayout`].
```
I agree. I've also generally cleaned up / added comments.

nicoburns commented:
I believe I've addressed all of the review feedback, so this is ready for re-review. I've also made #146 independent of this PR if anybody has time to review that (which is hopefully a little less controversial than this one).
